Init button

ianhi opened the pull request:
Closes: https://github.com/jupyterlab/jupyterlab-git/issues/589

New UI:
![image](https://user-images.githubusercontent.com/10111092/79645238-12d38000-817c-11ea-9a3b-37f6a7e665d4.png)

I think this still needs some work on where to put and how to style the file path. The two big issues are:
1. If you are in the jupyter root directory the file path looks weird:
![image](https://user-images.githubusercontent.com/10111092/79645265-3ac2e380-817c-11ea-8614-bbb32a2cf5ce.png)
I didn't see an easy way to get the absolute file path from the filebrowser in order to get the folder name in this case.

2. A really long folder name renders poorly
![image](https://user-images.githubusercontent.com/10111092/79645293-752c8080-817c-11ea-85a5-9f274727e38e.png)


I briefly explored the suggestion of replacing the Current Repository header with the name of the current repo but didn't see a super easy way to do this. 

### Better styling ideas:

Maybe just don't list the name of the current folder. i.e. say: "You aren't currently in a git repository you can either ...."

Or: style the filepath text differently and limit its length.

### Slow refresh:
I also found that sometimes the filepath can be slow to update:
![slow](https://user-images.githubusercontent.com/10111092/79651606-6eebd380-817f-11ea-9956-bebe17e09ba2.gif)
![slow2](https://user-images.githubusercontent.com/10111092/79651610-6f846a00-817f-11ea-99b2-127db9b21447.gif)

### Adding a clone button
If https://github.com/jupyterlab/jupyterlab-git/pull/596 gets merged then it would be nice to add a third button for cloning to this interface.

kgryte commented:
If I had to hazard a guess as to why things might be "slow" is because of the poll interval. Try adjusting the poll interval in the extension settings and seeing if you see the same lag.

ianhi commented:
> " is because of the poll interval

Yup! that was definitely it. In the above gifs it updates every 3 seconds, which is the default polling rate.

I had a further thought about how to display the folder names.
1. For the jupyter root dir just have prepared text:
![image](https://user-images.githubusercontent.com/10111092/79706726-397edd00-8288-11ea-96de-78f97b357d1c.png)

2. For folders over than the jupyter root level just display the final folder in the path in bold:
e.g. here Documents/van_kampen becomes just van_kampen
![image](https://user-images.githubusercontent.com/10111092/79706712-2ff57500-8288-11ea-9f93-dc46da66d1f6.png)

fcollonval commented:
@ianhi Your latest proposal looks nice. You could even add a `title` on the folder name element to display the full path when hovered.

ianhi commented:
Good suggestion with the title, I added that. I also made all three of the buttons be the same width because otherwise they were slightly misaligned.

ianhi commented:
@fcollonval you had suggested removing the top part of the UI
![image](https://user-images.githubusercontent.com/10111092/79818237-4fa69f00-8355-11ea-9976-4511647a8259.png)
when not in a git repo. I think this makes sense was a pretty easy fix. But, to me it makes the text feel almost too high up?
![image](https://user-images.githubusercontent.com/10111092/79818205-3dc4fc00-8355-11ea-83ae-3612cde8cb67.png)

Also, looking at the test failures they don't seem to be related to these changes:
https://travis-ci.org/github/jupyterlab/jupyterlab-git/builds/677526123#L804
```bash
pluggy.manager.PluginValidationError: Plugin 'asyncio' could not be loaded: (pytest 4.3.1 (/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages), Requirement.parse('pytest>=5.4.0'))!

The command "pytest jupyterlab_git -r ap" exited with 1.
```

ianhi commented:
Also this PR will make the gif in the readme slightly out of date.

telamonian commented:
> Also this PR will make the gif in the readme slightly out of date.

Well, if you feel bad about it you can always record a new one :)

I used the free/oss app LICEcap to make the existing animation, if you actually do want to take a stab at it.

@ianhi I like the direction this PR is taking. I just pushed a fix, the tests are now passing

fcollonval commented:
@ianhi Thanks for the great work. This looks quite good. I think the vertical space is find.

The only change, I would suggest, is to left align the text paragraph. This feels more natural for large text.

ianhi commented:
I'll add the left align. I also need to resolve the conflicts, I think rebasing is the appropriate thing here?

I'd also like to add icons on the left side of the buttons (maybe inside the buttons?). Maybe a folder icon for the filebrowser, a plus sign for init, and the pull icon for clone. I think that after the first time you've seen the dialog and know what to expect the icons will make the options easier to visually process.

fcollonval commented:
> I think rebasing is the appropriate thing here?

Yep - but be aware that the port to JupyterLab 2.0 has been merged. So this means you will need to install JLab 2.0 in your dev environment, reinstall the extension in it and rebuild the frontend.

> I'd also like to add icons on the left side of the buttons (maybe inside the buttons?)

Inside the button makes more sense.

ianhi commented:
Rebased on master and squashed some of the commits together. New UI looks like:
![image](https://user-images.githubusercontent.com/10111092/80033381-9bbf2400-84ba-11ea-82dc-0459064f6b6e.png).

It feels as though I'm missing some punctuation in the second sentence, but proper comma placement has always been a weak point for me so I'm not quite sure.


I tried to add icons using `<LabIcon.resolveReact icon='git-pull' />` but wasn't able to scale the icon and keep the icon on the same line as the rest of the text in the button. For example:
![sad](https://user-images.githubusercontent.com/10111092/80033458-c6a97800-84ba-11ea-98d3-49e0744dce55.png)
But I won't lose sleep over the lack of an icon.

kgryte commented on src/components/GitPanel.tsx:
```diff
@@ -295,16 +302,42 @@ export class GitPanel extends React.Component<
+          ) : (
+            'You are not currently in'
+          )}{' '}
+          a git repository. To use git navigate to a local repository, intialize
```
```suggestion
          a Git repository. To use Git, navigate to a local repository, initialize
```

ianhi commented:
amended to fix the lint error. 

Unless anyone has a suggestion on how to properly incorporate icons inside the buttons then I'm happy with the way this looks.

fcollonval commented:
@meeseeksdev backport to 0.11.x

ianhi commented on src/components/GitPanel.tsx:
```diff
@@ -295,16 +302,42 @@ export class GitPanel extends React.Component<
+        <br />
+        <button
+          className={repoButtonClass}
+          onClick={() => this.props.model.commands.execute('git:init')}
```
Thanks for leaving these comments. Very helpful as I learn more!
